.github/workflows/test-lang-rust-ci.yml

```diff
@@ -58,20 +58,59 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v4
 
-      - name: Cache Cargo
+      - name: Cache Cargo for 1.65.0
+        if: matrix.rust == '1.65.0' && matrix.target == 'x86_64-unknown-linux-gnu'
         uses: actions/cache@v3
         with:
           # these represent dependencies downloaded by cargo
           # and thus do not depend on the OS, arch nor rust version.
           path: ~/.cargo
-          key: cargo-cache1-
+          key: cargo-cache1-1.65.0-${{ hashFiles('**/Cargo.lock') }}
+
+      - name: Cache Cargo for 1.65.0 (Restore Only)
+        if: matrix.rust == '1.65.0' && matrix.target != 'x86_64-unknown-linux-gnu'
+        uses: actions/cache/restore@v3
```

Comment:
This looks weird. 
` actions/cache@v3` does two things - first it restores the cache and later in the post-processing saves the cache.
Now you introduce the new `actions/cache/restore@v3` which gives you more control when to `restore` but it should be used with `actions/cache/save@v3` to have a better control **when** to `save`.

Reply:
Having only `restore` for some cache is intended.
For `lang/rust/target`, if a target is `wasm32-unknown-unknown`, cargo test doesn't run. So if `wasm32-unknown-unknown` finishes earlier than `x86_64-unknown-linux-gnu`, the dependencies built for tests will not be cached right?

For `~/.cargo`, it might not be necessary to separate by conditions.

Reply:
I am still very confused about these changes :-/
IMO we should add `${{ matrix.target }}` to the cache key. This way x86_64 caches won't interfere with the wasm ones.

Reply:
Also the handling of pre-1.70 and post-1.70 adds to the confusion.
Let's just disable the new sparse protocol until MSRV is 1.70.0 via https://doc.rust-lang.org/cargo/reference/config.html#registriescrates-ioprotocol (i.e. a global env variable for the whole workflow)

Reply:
> I am still very confused about these changes :-/
IMO we should add ${{ matrix.target }} to the cache key. This way x86_64 caches won't interfere with the wasm ones.

As I explained [here](https://github.com/apache/avro/pull/2538#discussion_r1344376019), I didn't intentionally use `${{ matrix.target }}`.

> Also the handling of pre-1.70 and post-1.70 adds to the confusion.
Let's just disable the new sparse protocol until MSRV is 1.70.0 via https://doc.rust-lang.org/cargo/reference/config.html#registriescrates-ioprotocol (i.e. a global env variable for the whole workflow)

This idea seems awesome!